File: mobile/src/scraper/youtube.ts
import AsyncStorage from '@react-native-async-storage/async-storage'
import type { FeedItem } from '../localDb'
import { DESKTOP_UA, now } from './utils'
import { fetchYouTube } from '../connectors/youtube'

function unescape(s: string): string {
  return s
    .replace(/\\n/g, ' ')
    .replace(/\\"/g, '"')
    .replace(/\\'/g, "'")
    .replace(/\\\\/g, '\\')
    .replace(/\\u([\da-fA-F]{4})/g, (_, h) => String.fromCharCode(parseInt(h, 16)))
    .replace(/ - YouTube$/, '')
    .trim()
}

// Parse YouTube's relative-time strings ("3 months ago", "2週間前", etc.) into ISO dates.
function parseYouTubeDate(seg: string): string {
  const raw =
    seg.match(/"publishedTimeText":\{"simpleText":"([^"]+)"/)?.[1] ??
    seg.match(/"publishedTimeText":\{"runs":\[\{"text":"([^"]+)"/)?.[1] ??
    ''
  if (!raw) return now()
  const m = raw.match(/(\d+)\s*(second|minute|hour|day|week|month|year|秒|分|時間|日|週間?|ヶ月|年)/i)
  if (!m) return now()
  const n = parseInt(m[1], 10)
  const u = m[2].toLowerCase()
  let ms = 0
  if (u.startsWith('sec') || u === '秒') ms = n * 1_000
  else if (u.startsWith('min') || u === '分') ms = n * 60_000
  else if (u.startsWith('hour') || u === '時間') ms = n * 3_600_000
  else if (u.startsWith('day') || u === '日') ms = n * 86_400_000
  else if (u.startsWith('week') || u === '週') ms = n * 7 * 86_400_000
  else if (u.startsWith('month') || u === 'ヶ月') ms = n * 30 * 86_400_000
  else if (u.startsWith('year') || u === '年') ms = n * 365 * 86_400_000
  return ms > 0 ? new Date(Date.now() - ms).toISOString() : now()
}

export async function scrapeYouTube(keyword: string, mode: string): Promise<FeedItem[]> {
  // Use YouTube Data API if the user has configured a key
  let apiKey = ''
  try {
    apiKey = (await AsyncStorage.getItem('@otterpia/yt_key'))?.trim() ?? ''
  } catch {
    apiKey = ''
  }
  if (apiKey) {
    try {
      const items = await fetchYouTube(keyword, mode, apiKey)
      if (items.length > 0) {
        return items.map(item => ({
          id: `youtube:${item.item_id}`,
          platform: 'youtube',
          url: item.url,
          title: item.title ?? null,
          content_text: item.content_text ?? null,
          author: item.author ?? null,
          thumbnail_url: item.thumbnail_url ?? null,
          media_type: item.media_type,
          published_at: item.published_at.toISOString(),
          watch_term_keyword: keyword,
          fetched_at: now(),
        }))
      }
    } catch (e) {
      console.log('[youtube:api] failed, falling back to scrape:', e)
    }
  }

  const results: FeedItem[] = []
  const seen = new Set<string>()
  try {
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), 15_000)
    let html: string
    try {
      const res = await fetch(
        `https://www.youtube.com/results?search_query=${encodeURIComponent(keyword)}`,
        {
          signal: controller.signal,
          headers: {
            'User-Agent': DESKTOP_UA,
            'Accept-Language': 'ja,en;q=0.9',
            'Accept': 'text/html,*/*',
          },
        },
      )
      if (!res.ok) throw new Error(`HTTP ${res.status}`)
      html = await res.text()
    } finally {
      clearTimeout(timer)
    }

    console.log(`[youtube] html=${html.length}`)

    // Anchor on "videoRenderer":{"videoId":"..." — this is always the start of a
    // real video card, unlike bare "videoId" which appears dozens of times per card
    // (in watch endpoints, thumbnails, related items, etc.)
    const re = /"videoRenderer":\{"videoId":"([A-Za-z0-9_-]{11})"/g
    let m: RegExpExecArray | null
    while (results.length < 20 && (m = re.exec(html)) !== null) {
      const vid = m[1]
      if (seen.has(vid)) continue
      seen.add(vid)

      // Title is always in the first ~800 chars of the videoRenderer object.
      // Key fix: don't require "}" after the captured text — YouTube's JSON has
      // more fields after "text":"TITLE" so it ends with "," not "}".
      const seg = html.slice(m.index, Math.min(html.length, m.index + 1200))

      let rawTitle: string | null =
        // Primary: title.runs[0].text
        seg.match(/"title":\{"runs":\[\{"text":"((?:[^"\\]|\\.)*)"/)
          ?.[1] ??
        // Fallback: simpleText (used in some layouts)
        seg.match(/"title":\{"simpleText":"((?:[^"\\]|\\.)*)"/)
          ?.[1] ??
        // Fallback: accessibility label attached to title (strip channel/stats suffix)
        seg.match(/"title":\{"accessibility":\{"accessibilityData":\{"label":"((?:[^"\\]|\\.)*)"/)
          ?.[1]
          ?.replace(/ - (?:by |\d[\d,.]*\s*(?:回|views?|再生)).*/i, '')
          .trim() ??
        null

      results.push({
        id: `youtube:${vid}`,
        platform: 'youtube',
        url: `https://www.youtube.com/watch?v=${vid}`,
        title: rawTitle ? unescape(rawTitle) : null,
        content_text: null, author: null, thumbnail_url: null,
        media_type: 'video', published_at: parseYouTubeDate(seg),
        watch_term_keyword: keyword, fetched_at: now(),
      })
    }

    console.log(`[youtube] found=${results.length}`)
  } catch (e) {
    console.log(`[youtube] failed:`, e)
  }
  return results
}

File: backend/app/connectors/youtube.py
import json
import logging
import re
from datetime import datetime, timedelta, timezone
from typing import Optional

import httpx

from app.connectors.base import BaseConnector, SourceItemCreate

log = logging.getLogger(__name__)


def _parse_youtube_relative(text: str) -> Optional[datetime]:
    """Convert YouTube relative timestamps ('2 days ago', '3ヶ月前') to UTC datetimes."""
    if not text:
        return None
    now = datetime.now(timezone.utc)
    m = re.search(r'(\d+)[\s 　]*(second|minute|hour|day|week|month|year|秒|分|時間|日|週間?|週|ヶ月|か月|年)', text.lower())
    if not m:
        return None
    n, unit = int(m.group(1)), m.group(2)
    if unit in ('second', '秒'):      return now - timedelta(seconds=n)
    if unit in ('minute', '分'):      return now - timedelta(minutes=n)
    if unit in ('hour', '時間'):      return now - timedelta(hours=n)
    if unit in ('day', '日'):         return now - timedelta(days=n)
    if unit in ('week', '週', '週間'): return now - timedelta(weeks=n)
    if unit in ('month', 'ヶ月', 'か月'): return now - timedelta(days=n * 30)
    if unit in ('year', '年'):        return now - timedelta(days=n * 365)
    return None


class YouTubeConnector(BaseConnector):
    PLATFORM = "youtube"
    SUPPORTS_MEDIA_FILTER = True

    _SEARCH_URL = "https://www.googleapis.com/youtube/v3/search"

    def __init__(self, api_key: str) -> None:
        self.api_key = api_key

    async def _fetch_api(self, keyword: str) -> list[SourceItemCreate]:
        cutoff = datetime.now(timezone.utc) - timedelta(days=90)
        params = {
            "part": "snippet",
            "q": keyword,
            "type": "video",
            "order": "date",
            "publishedAfter": cutoff.strftime("%Y-%m-%dT%H:%M:%SZ"),
            "maxResults": 25,
            "key": self.api_key,
        }

        async with httpx.AsyncClient(timeout=10.0) as client:
            resp = await client.get(self._SEARCH_URL, params=params)
            resp.raise_for_status()
            data = resp.json()

        items: list[SourceItemCreate] = []
        for raw in data.get("items", []):
            vid_id = raw["id"].get("videoId")
            if not vid_id:
                continue
            snippet = raw["snippet"]
            published = datetime.fromisoformat(snippet["publishedAt"].replace("Z", "+00:00"))
            thumb = snippet.get("thumbnails", {}).get("medium", {}).get("url")
            items.append(
                SourceItemCreate(
                    platform=self.PLATFORM,
                    item_id=vid_id,
                    url=f"https://www.youtube.com/watch?v={vid_id}",
                    published_at=published,
                    media_type="video",
                    author=snippet.get("channelTitle"),
                    title=snippet.get("title"),
                    content_text=snippet.get("description"),
                    thumbnail_url=thumb,
                    raw_payload=raw,
                )
            )

        return items

    async def _fetch_scrape(self, keyword: str) -> list[SourceItemCreate]:
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
            "Accept-Language": "ja,ja-JP;q=0.9,en;q=0.8",
        }
        url = f"https://www.youtube.com/results?search_query={keyword}"

        async with httpx.AsyncClient(timeout=15.0, headers=headers, follow_redirects=True) as client:
            resp = await client.get(url)
            if not resp.is_success:
                log.warning("YouTube search scrape failed with status code %d", resp.status_code)
                return []

        # Find ytInitialData JSON inside HTML
        m = re.search(r"ytInitialData\s*=\s*({.+?});", resp.text, re.S)
        if not m:
            log.warning("ytInitialData not found in YouTube search scrape response")
            return []

        try:
            data = json.loads(m.group(1))
            contents = (
                data.get("contents", {})
                .get("twoColumnSearchResultsRenderer", {})
                .get("primaryContents", {})
                .get("sectionListRenderer", {})
                .get("contents", [])
            )
        except Exception as e:
            log.warning("Failed to parse ytInitialData JSON: %s", e)
            return []

        cutoff = datetime.now(timezone.utc) - timedelta(days=90)
        items: list[SourceItemCreate] = []
        for content in contents:
            item_section = content.get("itemSectionRenderer", {})
            for item in item_section.get("contents", []):
                if "videoRenderer" in item:
                    vr = item["videoRenderer"]
                    vid_id = vr.get("videoId")
                    if not vid_id:
                        continue

                    title = vr.get("title", {}).get("runs", [{}])[0].get("text")
                    channel = vr.get("ownerText", {}).get("runs", [{}])[0].get("text")

                    desc = ""
                    snippets = vr.get("detailedMetadataSnippets", [])
                    if snippets:
                        desc = snippets[0].get("snippetText", {}).get("runs", [{}])[0].get("text", "")

                    thumb = None
                    thumbnails = vr.get("thumbnail", {}).get("thumbnails", [])
                    if thumbnails:
                        thumb = thumbnails[0].get("url")

                    rel_text = vr.get("publishedTimeText", {}).get("simpleText", "")
                    published_at = _parse_youtube_relative(rel_text) or datetime.now(timezone.utc)

                    if published_at < cutoff:
                        continue

                    items.append(
                        SourceItemCreate(
                            platform=self.PLATFORM,
                            item_id=str(vid_id),
                            url=f"https://www.youtube.com/watch?v={vid_id}",
                            published_at=published_at,
                            media_type="video",
                            author=str(channel) if channel else None,
                            title=str(title) if title else None,
                            content_text=str(desc) if desc else None,
                            thumbnail_url=thumb,
                            raw_payload=item,
                        )
                    )

        return items

    async def fetch(self, keyword: str, mode: str) -> list[SourceItemCreate]:
        # Always try API first if credential is provided
        if self.api_key:
            try:
                return await self._fetch_api(keyword)
            except Exception as e:
                log.warning("YouTube API fetch failed, falling back to scrape. Error: %s", e)

        # Scrape fallback
        try:
            return await self._fetch_scrape(keyword)
        except Exception as e:
            log.warning("YouTube scrape fetch failed. Error: %s", e)
            return []
